### PR TITLE
Support triton kernels v3.4.0 for fused_moe

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation 

This PR is to follow up https://github.com/sgl-project/sglang/pull/7966 [1/N] MoE Refactor: refactor select_experts. It supports triton_kernels v3.4.0 for fused moe.
This PR involves a change to the TopKOutput data structure. 

Co-authored-by: @ch-wan

Note: it requires pytorch-triton.
```
Name: pytorch-triton
Version: 3.4.0+gitae848267
```

The result is as expected.
```
➜  python git:(support_triton_moe) ✗ python3 -m sglang.launch_server --model Qwen/Qwen3-30B-A3B --tp-size 8 --port 30000 --enable-triton-kernel-moe
......
[2025-07-22 06:17:55 TP0] Capture cuda graph begin. This can take up to several minutes. avail mem=30.36 GB
[2025-07-22 06:17:55 TP0] Capture cuda graph bs [1, 2, 4, 8, 16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96, 104, 112, 120, 128, 136, 144, 152, 160, 168, 176, 184, 192, 200, 208, 216, 224, 232, 240, 248, 256, 272, 288, 304, 320, 336, 352, 368, 384, 400, 416, 432, 448, 464, 480, 496, 512]
Capturing batches (bs=1 avail_mem=23.02 GB):  98%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊   | 50/51 [00:36<00:00,  1.45it/s][2025-07-22 06:18:33 TP5] Registering 4947 cuda graph addresses
Capturing batches (bs=1 avail_mem=23.02 GB): 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 51/51 [00:37<00:00,  1.35it/s]
[2025-07-22 06:18:33 TP1] Registering 4947 cuda graph addresses
[2025-07-22 06:18:33 TP3] Registering 4947 cuda graph addresses
[2025-07-22 06:18:33 TP7] Registering 4947 cuda graph addresses
[2025-07-22 06:18:33 TP6] Registering 4947 cuda graph addresses
[2025-07-22 06:18:33 TP4] Registering 4947 cuda graph addresses
[2025-07-22 06:18:33 TP2] Registering 4947 cuda graph addresses
[2025-07-22 06:18:33 TP0] Registering 4947 cuda graph addresses
[2025-07-22 06:18:33 TP0] Capture cuda graph end. Time elapsed: 38.61 s. mem usage=7.36 GB. avail mem=23.00 GB.
[2025-07-22 06:18:34 TP0] max_total_num_tokens=6004242, chunked_prefill_size=16384, max_prefill_tokens=16384, max_running_requests=4096, context_len=40960, available_gpu_mem=23.00 GB
[2025-07-22 06:18:35] INFO:     Started server process [163598]
[2025-07-22 06:18:35] INFO:     Waiting for application startup.
[2025-07-22 06:18:35] INFO:     Application startup complete.
[2025-07-22 06:18:35] INFO:     Uvicorn running on http://127.0.0.1:30000 (Press CTRL+C to quit)
[2025-07-22 06:18:36] INFO:     127.0.0.1:42888 - "GET /get_model_info HTTP/1.1" 200 OK
[2025-07-22 06:18:36 TP0] Prefill batch. #new-seq: 1, #new-token: 6, #cached-token: 0, token usage: 0.00, #running-req: 0, #queue-req: 0,
[2025-07-22 06:18:42] INFO:     127.0.0.1:42892 - "POST /generate HTTP/1.1" 200 OK
[2025-07-22 06:18:42] The server is fired up and ready to roll!
[2025-07-22 06:18:52 TP0] Prefill batch. #new-seq: 1, #new-token: 33, #cached-token: 0, token usage: 0.00, #running-req: 0, #queue-req: 0,
[2025-07-22 06:18:53 TP0] Decode batch. #running-req: 1, #token: 66, token usage: 0.00, cuda graph: True, gen throughput (token/s): 2.14, #queue-req: 0,
[2025-07-22 06:18:53 TP0] Decode batch. #running-req: 1, #token: 106, token usage: 0.00, cuda graph: True, gen throughput (token/s): 201.36, #queue-req: 0,
[2025-07-22 06:18:53 TP0] Decode batch. #running-req: 1, #token: 146, token usage: 0.00, cuda graph: True, gen throughput (token/s): 194.93, #queue-req: 0,
[2025-07-22 06:18:53 TP0] Decode batch. #running-req: 1, #token: 186, token usage: 0.00, cuda graph: True, gen throughput (token/s): 191.74, #queue-req: 0,
[2025-07-22 06:18:53 TP0] Decode batch. #running-req: 1, #token: 226, token usage: 0.00, cuda graph: True, gen throughput (token/s): 192.56, #queue-req: 0,

➜  /sgl-workspace python test_openai.py
ChatCompletion(id='aab4b61556c84cb292224b0408ff4b39', choices=[Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content="<think>\nOkay, the user asked for three countries and their capitals, and then wants to know how I rank them. Let me think about this. First, I need to provide the list of countries and their capitals. But the user didn't specify which countries, so I should choose three that are commonly known. Maybe the United States, the United Kingdom, and France. But wait, the United Kingdom is a country, but its capital is London. The United States is a country, and its capital is Washington D.C. France is a country, and its capital is Paris. So that's three countries. Then, the user wants to know how I rank them. But what does the user mean by ranking? Maybe the user wants to know the order of the countries, or the order of the capitals. Or maybe the user wants to know the ranking of the countries based on some criteria, like population, or size, or something else. But the user didn't specify. So I need", refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None, reasoning_content=None), matched_stop=None)], created=1753190389, model='default', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=200, prompt_tokens=33, total_tokens=233, completion_tokens_details=None, prompt_tokens_details=None))
```

## Modifications

Modified qwen3_moe for the moment, other MoE models adapt are WIP.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/quantization/unquant.py [hunk 1]
```diff
@@ -219,16 +227,17 @@ def forward_cuda(
...
-            # )
+            topk_weights, _, router_logits = topk_output
+            top_k = topk_output.top_k
+            renormalize = topk_output.renormalize
+            return self.triton_kernel_moe_forward(
+                hidden_states=x,
+                w1=layer.w13_weight,
+                w2=layer.w2_weight,
+                gating_output=router_logits,
+                topk=top_k,
```
Comment (BBuf) [inline python/sglang/srt/layers/quantization/unquant.py]:
```suggestion
                topk=topk_output.top_k,
                renormalize = topk_output.renormalize
```

Comment (yuan-luo) [reply python/sglang/srt/layers/quantization/unquant.py]: Revised.

**Comment by yuan-luo** [comment]
We can see cuda graph capture error in current CI results. [07/24] 
The reason is after modifying TopKOutput to the following, 
```
@dataclass
class TopKOutput:
    topk_weights: torch.Tensor
    topk_ids: torch.Tensor
    router_logits: torch.Tensor
    top_k: int
    renormalize: bool

    def __iter__(self):
        yield self.topk_weights
        yield self.topk_ids
        yield self.router_logits
```
dynamic memory allocation may occur due to changes in the __iter__ method. Previously, a static tuple of tensors was returned, which allowed the compiler to capture them in advance. However, after modifying the data class, it now returns a dynamic container, which results in shape mismatches during instance passing of the data class. This issue becomes especially pronounced when the top_k parameter is changed, leading to larger shape differences and memory allocation errors. The root cause lies in the discrepancy between the shapes captured during the tracing phase and those during actual execution.

It may still be necessary to revert to the original implementation as the following. 
```
class TopKOutput(NamedTuple):
    topk_weights: torch.Tensor
    topk_ids: torch.Tensor
    router_logits: torch.Tensor
    top_k: int
    renormalize: bool
```

**Review by ch-wan — APPROVED** [review]

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 2]
```diff
@@ -302,7 +302,7 @@ def forward_deepgemm(
         hidden_states_dtype = hidden_states.dtype
         hidden_states_device = hidden_states.device
 
-        topk_weights, topk_ids, _ = topk_output
+        topk_weights, topk_ids, _, _, _ = topk_output
```
Comment (merrymercy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: this seems to fragile.

1. How do you make sure you change all places?
2. Is it better to use the name to index the variables

Comment (yuan-luo) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: @merrymercy , thanks for the comments. The current approach is the agreement @ch-wan and I discussed and reached. The background is that (@ch-wan proposed) we plan to do another round non-trivial refactor to replace all the TopK with a concrete TopK class such as FBGemmTopK, TritonTopK, FlashInferTopK etc. We have gone through all the places of the TopKOutput and made sure the change covers every place. 
Given this, shall we keep the current implementation and push the PR forward?

Comment (ch-wan) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: Yes, this is a part of my on-going refactor. We temporarily use this format because oai needs triton kernel support.
